File: packages/identity/go/jwks.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// JWKS fetcher with zone-scoped, 5-min in-memory cache.

package identity

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"net"
	"net/http"
	"net/url"
	"os"
	"sync"
	"time"
)

const (
	jwksTTL          = 5 * time.Minute
	jwksFetchTimeout = 10 * time.Second
)

type jwksEntry struct {
	keys      map[string]*ecdsa.PublicKey
	fetchedAt time.Time
}

var (
	jwksMu     sync.RWMutex
	jwksCache  = map[string]*jwksEntry{}
	jwksClient = &http.Client{Timeout: jwksFetchTimeout}
)

// GetJWKS returns the cached key set for one zone of the issuer, fetching if
// missing or stale. STS serves one signing keyset per zone, so zoneID is
// required.
func GetJWKS(issuer, zoneID string) (map[string]*ecdsa.PublicKey, error) {
	return GetJWKSContext(context.Background(), issuer, zoneID)
}

// GetJWKSContext is GetJWKS with caller-supplied cancellation.
func GetJWKSContext(ctx context.Context, issuer, zoneID string) (map[string]*ecdsa.PublicKey, error) {
	if err := assertSecureIssuer(issuer); err != nil {
		return nil, err
	}
	if zoneID == "" {
		return nil, fmt.Errorf("zone_id required: STS serves one signing keyset per zone")
	}
	fetchURL := issuer + "/.well-known/jwks.json?" + url.Values{"zone_id": {zoneID}}.Encode()
	cacheKey := issuer + "\x00" + zoneID

	jwksMu.RLock()
	entry, ok := jwksCache[cacheKey]
	jwksMu.RUnlock()
	if ok && time.Since(entry.fetchedAt) < jwksTTL {
		return entry.keys, nil
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fetchURL, nil)
	if err != nil {
		return nil, err
	}
	resp, err := jwksClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("jwks fetch %d", resp.StatusCode)
	}

	var body struct {
		Keys []json.RawMessage `json:"keys"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return nil, err
	}

	keys := make(map[string]*ecdsa.PublicKey, len(body.Keys))
	for _, raw := range body.Keys {
		key, kid, err := ParseECJWK(raw)
		if err != nil {
			continue
		}
		keys[kid] = key
	}

	jwksMu.Lock()
	jwksCache[cacheKey] = &jwksEntry{keys: keys, fetchedAt: time.Now()}
	jwksMu.Unlock()
	return keys, nil
}

// ResetJWKSCache clears the in-memory JWKS cache. Intended for tests.
func ResetJWKSCache() {
	jwksMu.Lock()
	jwksCache = map[string]*jwksEntry{}
	jwksMu.Unlock()
}

// assertSecureIssuer requires the issuer to use https. http is permitted only for
// loopback hosts, or for any host when CARACAL_ALLOW_INSECURE_CONFIG_URLS=true, so
// local development and trusted-network deployments are not blocked.
func assertSecureIssuer(issuer string) error {
	parsed, err := url.Parse(issuer)
	if err != nil {
		return fmt.Errorf("invalid issuer url: %w", err)
	}
	switch parsed.Scheme {
	case "https":
		return nil
	case "http":
		if isLoopbackHost(parsed.Hostname()) || os.Getenv("CARACAL_ALLOW_INSECURE_CONFIG_URLS") == "true" {
			return nil
		}
		return fmt.Errorf("insecure issuer scheme: http requires a loopback host or CARACAL_ALLOW_INSECURE_CONFIG_URLS=true")
	default:
		return fmt.Errorf("unsupported issuer scheme: %q", parsed.Scheme)
	}
}

func isLoopbackHost(host string) bool {
	if host == "localhost" {
		return true
	}
	ip := net.ParseIP(host)
	return ip != nil && ip.IsLoopback()
}

// ParseECJWK decodes a single EC P-256 JWK (RFC 7517) and returns the public key together
// with its key id. Returns an error when the key type, curve, or kid is missing/unsupported,
// or when the X/Y coordinates fail base64url decoding.
func ParseECJWK(raw json.RawMessage) (*ecdsa.PublicKey, string, error) {
	var key struct {
		Kty    string   `json:"kty"`
		Crv    string   `json:"crv"`
		Kid    string   `json:"kid"`
		Alg    string   `json:"alg"`
		Use    string   `json:"use"`
		KeyOps []string `json:"key_ops"`
		X      string   `json:"x"`
		Y      string   `json:"y"`
	}
	if err := json.Unmarshal(raw, &key); err != nil {
		return nil, "", err
	}
	if key.Kty != "EC" || key.Crv != "P-256" || key.Kid == "" || key.Alg != "ES256" || (key.Use != "" && key.Use != "sig") {
		return nil, "", fmt.Errorf("unsupported jwk")
	}
	if len(key.KeyOps) > 0 && !containsKeyOp(key.KeyOps, "verify") {
		return nil, "", fmt.Errorf("unsupported jwk")
	}
	xBytes, err := base64.RawURLEncoding.DecodeString(key.X)
	if err != nil {
		return nil, "", err
	}
	yBytes, err := base64.RawURLEncoding.DecodeString(key.Y)
	if err != nil {
		return nil, "", err
	}
	if len(xBytes) != 32 || len(yBytes) != 32 {
		return nil, "", fmt.Errorf("invalid jwk coordinates")
	}
	curve := elliptic.P256()
	x := new(big.Int).SetBytes(xBytes)
	y := new(big.Int).SetBytes(yBytes)
	if !curve.IsOnCurve(x, y) {
		return nil, "", fmt.Errorf("invalid jwk point")
	}
	return &ecdsa.PublicKey{Curve: curve, X: x, Y: y}, key.Kid, nil
}

func containsKeyOp(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

File: packages/identity/ts/src/jwks.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// JWKS cache for mandate signature verification.

import { createLocalJWKSet, type JWTVerifyGetKey } from 'jose'

interface JwksEntry {
  keySet: JWTVerifyGetKey
  fetchedAt: number
  revalidating?: Promise<JWTVerifyGetKey>
}

export interface JwksCacheOptions {
  ttlMs?: number
  fetchTimeoutMs?: number
  fetchImpl?: typeof fetch
}

export interface JwksCache {
  getKeySet: (issuer: string) => Promise<JWTVerifyGetKey>
  warm: (issuer: string) => Promise<void>
  clear: (issuer?: string) => void
}

const DEFAULT_CACHE_TTL_MS = 5 * 60 * 1000
const DEFAULT_FETCH_TIMEOUT_MS = 5_000

function assertSecureIssuer(issuer: string): void {
  const parsed = new URL(issuer)
  if (parsed.protocol === 'https:') return
  if (parsed.protocol === 'http:') {
    const insecureAllowed =
      isLoopbackHost(parsed.hostname) ||
      (process.env.NODE_ENV ?? 'development') === 'development' ||
      process.env.CARACAL_ALLOW_INSECURE_CONFIG_URLS === 'true'
    if (insecureAllowed) return
    throw new Error('insecure issuer scheme: http requires a loopback host or development mode')
  }
  throw new Error(`unsupported issuer scheme: ${parsed.protocol}`)
}

function isLoopbackHost(host: string): boolean {
  if (host === 'localhost') return true
  if (host === '::1' || host === '[::1]') return true
  return /^127(?:\.\d{1,3}){3}$/.test(host)
}

export function createJwksCache(opts: JwksCacheOptions = {}): JwksCache {
  const ttlMs = opts.ttlMs ?? DEFAULT_CACHE_TTL_MS
  const fetchTimeoutMs = opts.fetchTimeoutMs ?? DEFAULT_FETCH_TIMEOUT_MS
  const fetchImpl = opts.fetchImpl
  const cache = new Map<string, JwksEntry>()

  async function getKeySet(issuer: string): Promise<JWTVerifyGetKey> {
    const now = Date.now()
    const entry = cache.get(issuer)
    if (entry) {
      const age = now - entry.fetchedAt
      if (age < ttlMs) return entry.keySet
      if (!entry.revalidating) {
        entry.revalidating = fetchAndStore(issuer)
          .catch(() => entry.keySet)
          .finally(() => {
            const current = cache.get(issuer)
            if (current) current.revalidating = undefined
          })
      }
      return entry.keySet
    }
    return fetchAndStore(issuer)
  }

  async function fetchAndStore(issuer: string): Promise<JWTVerifyGetKey> {
    assertSecureIssuer(issuer)
    const url = `${issuer.replace(/\/$/, '')}/.well-known/jwks.json`
    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), fetchTimeoutMs)
    try {
      const res = await (fetchImpl ?? fetch)(url, { signal: controller.signal })
      if (!res.ok) throw new Error(`JWKS fetch failed: ${res.status}`)
      const body = (await res.json()) as { keys: object[] }
      const keySet = createLocalJWKSet({ keys: body.keys } as Parameters<typeof createLocalJWKSet>[0])
      cache.set(issuer, { keySet, fetchedAt: Date.now() })
      return keySet
    } finally {
      clearTimeout(timeout)
    }
  }

  return {
    getKeySet,
    async warm(issuer: string): Promise<void> {
      await getKeySet(issuer)
    },
    clear(issuer?: string): void {
      if (issuer) {
        cache.delete(issuer)
        return
      }
      cache.clear()
    },
  }
}

const defaultCache = createJwksCache()

export async function getKeySet(issuer: string): Promise<JWTVerifyGetKey> {
  return defaultCache.getKeySet(issuer)
}

export async function warmJwks(issuer: string): Promise<void> {
  await defaultCache.warm(issuer)
}

export function clearJwksCache(issuer?: string): void {
  defaultCache.clear(issuer)
}
